src/wms/src/main/java/org/geoserver/wms/map/StyleQueryUtil.java

```diff
@@ -0,0 +1,452 @@
+/* (c) 2024 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.wms.map;
+
+import static org.geoserver.wms.decoration.MapDecorationLayout.FF;
+
+import java.awt.Graphics2D;
+import java.awt.Rectangle;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.geoserver.wms.WMSMapContent;
+import org.geotools.api.data.FeatureSource;
+import org.geotools.api.data.Query;
+import org.geotools.api.feature.type.FeatureType;
+import org.geotools.api.feature.type.GeometryDescriptor;
+import org.geotools.api.filter.Filter;
+import org.geotools.api.referencing.FactoryException;
+import org.geotools.api.referencing.crs.CoordinateReferenceSystem;
+import org.geotools.api.referencing.operation.TransformException;
+import org.geotools.api.style.FeatureTypeStyle;
+import org.geotools.api.style.Rule;
+import org.geotools.api.style.Style;
+import org.geotools.data.DataUtilities;
+import org.geotools.feature.FeatureTypes;
+import org.geotools.filter.IllegalFilterException;
+import org.geotools.filter.spatial.DefaultCRSFilterVisitor;
+import org.geotools.filter.spatial.ReprojectingFilterVisitor;
+import org.geotools.filter.visitor.SimplifyingFilterVisitor;
+import org.geotools.filter.visitor.SpatialFilterVisitor;
+import org.geotools.geometry.jts.Decimator;
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.geotools.map.Layer;
+import org.geotools.referencing.operation.transform.ProjectiveTransform;
+import org.geotools.renderer.lite.MetaBufferEstimator;
+import org.geotools.renderer.lite.RendererUtilities;
+import org.geotools.util.factory.Hints;
+import org.geotools.util.logging.Logging;
+
+/**
+ * Utility class for creating a query selecting those features relevant to the style and extent of
+ * the given map, for the given layer.
```

Comment:
I'd make it clear here too, that some of the same code is available buried somewhere down in GeoTools.
Also make it clear that unlike GeoTools, the query does not contain a selection of properties, because the usage points need to dump all properties (vector tiles use case) and indicate the current consumers are mapml and vector tiles.